README.md
# SDTTool
Version 0.9

*SDTTool* is a utility to read and convert *Smart Device Template* (*SDT*) definitions.

## Introduction
The [Smart Device Template](https://github.com/Homegateway/SmartDeviceTemplate) (SDT) schema defined by the Home Gateway Initiative (HGI) is a format to describe concrete and abstract devices as well as capabilities, data models and functions of those devices. See [SDT Introduction](https://github.com/Homegateway/SmartDeviceTemplate/blob/master/SDT/schema3.0/docs/Introduction.md) and [SDT Components](https://github.com/Homegateway/SmartDeviceTemplate/blob/master/SDT3.0/docs/SDT_Components.md) for an introduction to the SDT.

*SDTtool* is a Python 3 script that reads SDT files and generates output in various formats. It can be used to produce documentation for the Devices and ModuleClasses of an SDT as well as to convert between different versions of the SDT.


## Installation
### Prerequisites

#### Python 3.x
An installation of [Python 3](https://www.python.org/downloads/) (version 3.8 or higher) is required. 

#### Additional Packages
- SDTTool uses the template engine [Jinja2](http://jinja.pocoo.org) in version 3.0.3. You need to install it, e.g. via pip:

	pip install jinja2==3.0.3

- The *rich* package is used for nicer terminal output:

	pip install rich


### Copying
Download the latest release of *SDTTool* and copy them into a directory of your choice.

## Usage

*SDTTool* is run as follows:

	python3 sdttool.py -i anInputFile.xml 

This will read in the SDT version 2, the default, definitions from the file *anInputFile.xml* and writes the documentation in markdown format to standard out. Writing the output to a new file can be done like this:

	python3 sdttool.py -i anInputFile.xml -o anOutputFile.md

To change the output format the option ``-of`` must be given:

	python3 sdttool.py -i anInputFile.xml -o anOutputFile.opml -of opml

This would write the documentation in the OPML format to the file *anOutputFile.opml*.

The input format can be set by the option ``-if``.  

	python3 sdttool.py -i anInputFile.xml -o anOutputFile.opml -of opml -if sdt2

It is also possible to generate Java interfaces and classes from SDT version 3 with the ``java`` output format. In this case the ``-o`` argument must point to an output directory.

	python3 SDTTool.py -i anInputFile.xml -o anOutputDirectoy -of java

Skeletons of *definitions* and *paths* operations for a Swagger input file can be generated with the ``swagger`` output format. In this case the ``-o`` argument again must point to an output directory.

	python3 SDTTool.py -i anInputFile.xml -o anOutputDirectoy -of swagger

Running the script without with the ``-h`` option or without any argument will present a an overview about all possible command line parameters.

### Input Formats
- ``-if {sdt2,sdt3,sdt4}`` , ``--inputformat {sdt2,sdt3,sdt4}``: The input format to read. The default is *sdt4*.

The following input formats are supported for the ``-if`` or ``--inputformat`` command line argument:

- **sdt2**: SDT Version 2.0.1
- **sdt3**: SDT Version 3.0
- **sdt4**: SDT Version 4.0, the default

### Output Formats
Please note that not all output formats are available for every SDT version.

- ``-of {plain,opml,markdown,sdt3,sdt4,java,vorto-dsl,onem2m-svg,onem2m-xsd,swagger}``, ``--outputformat {plain,opml,markdown,sdt3,sdt4,java,vorto-dsl,onem2m-svg,onem2m-xsd,swagger}``: The output format for the result. The default is *markdown*.

Output formats for documentation for the ``-of`` or ``--outputformat`` command line argument:

- **plain**: This produces a plain text representation, with indentations, of the components of the input file.
- **markdown**: Markdown is a markup language with plain text formatting syntax designed so that it can be converted to many formats. GitHub natively supports markdown as its documentation format. See also the ``--markdowntables``argument below.
- **opml**: OPML (Outline Processor Markup Language) is a simple XML format for outlines, which can be imported in various mind mapping applications.
- **sdt3**: This output format is only valid when the input format is **sdt2**. It is used to convert SDT definitions from version 2 to version 3.
- **java**: This output format is only valid for the input format **sdt3**. It generates Java interfaces and classes for the input definition. For this output format the argument ``-o`` refers to an output directory, not a single file.
- **vorto-dsl**: This output format generates files that can be used to export Device, ModuleClass and data type definitions to an [Eclipse Vorto repository](http://vorto.eclipse.org). For this output format the argument ``-o`` refers to an output directory, not a single file.
- **onem2m-svg**: This output format generates SVG files that present the structure of Devices and ModuleClass in the graphical representation format used by [oneM2M](http://onem2m.org). For this output format the argument ``-o`` refers to an output directory, not a single file.
- **onem2m-xsd**: This output format generates XSD files according to the type definions of TS-0004 of the oneM2M specifications. For this output format the argument ``-o`` refers to an output directory, not a single file. The oneM2M domain needs to be specified via the ``--domain`` argument.   
In addition to the XSD files the following files are generated as well
	- Skeleton files for enum type definitions are generated in the *hd* sub-directory.
	- Files with newly found abbreviations (``_Abbreviations.*``). One file contains a python map, the other file contains the abbreviations in CSV format. Only new abbreviations, which are not found in the file that was specified with ``--abbreviationsinfile``, are added.
- **swagger**: This output format generates YAML files with Swagger definitions for ModuleClasses and Devices. These files cannot be used directly, but provide the necessary Swagger *definitions* skeletons and *paths* operations that can be copied and adapted to a real Swagger definition.  
For this output format the argument ``-o`` refers to an output directory, not a single file.

### Basic Arguments
- ``-i <filename>``, ``--infile <filename>``: Required argument. Specify the input file for the conversion.
- ``-o <filename>``, ``--outfile <filename>``: The output file or directory for the result. The default is stdout.

### oneM2M specific arguments
- ``--domain <domain name>``: Specify the domain name for XSD output.
- ``-ns <xsd prefix>``, ``--namespaceprefix <xsd prefix>``: Specify the XSD name space prefix for the model. This argument is mandatory when generating XSD and SVG drawings.
- ``--abbreviationsinfile <filename>``: Specify the file that contains a CSV table of already existing abbreviations.
- ``--abbreviationlength <integer>``: Specify the maximum length for abbreviations. The default is *5*.
- ``--xsdtargetnamespace <URI>`` : Specify the target namespace for the oneM2M XSD.
- ``-mv <version number>``, ``--modelversion <version number>`` : Specify the version of the model. This is used in the filenames of XSD and SVG files. "." characters are replaced with "_".

- ``--svg-with-attributes`` : Generate SVG drawings for ModuleClass attributes as well.



### Markdown Specific Arguments
- ``--markdowntables``: Generate tables instead of the usual list output style for markdown.
- ``--markdownpagebreak``: Insert page breaks before ModuleClasse and Device definitions.


### Other Arguments
- ``--hidedetails``: Hide the details of module classes and devices when generating the documentation.
- ``-lf <filename>``, ``--licensefile <filename>``: Add the text of the specified file as a license to the generated files.


### Configuration Files
Sometimes the number of command line arguments can get pretty big. Therefore, it is possible to put some or all arguments into a configuration file. This configuration file can be specified as follows

	python3 SDTTool.py @config

It is also possible to have more than one configuration file:

	python3 SDTTool.py @config1 @config2

or to mix command line arguments and configuration files:

	python3 SDTTool.py @config --markdowntables

## Limitations
- *SDTTool* does not validate the input XML. It is assumed that the input XML conforms to the SDT schema.
- Swagger support is not complete yet.

## Changelog

See the [Changelog](CHANGELOG.md) for all changes.

## Copyright and License
Copyright (c) 2015-2020 Deutsche Telekom AG

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

[http://www.apache.org/licenses/LICENSE-2.0](http://www.apache.org/licenses/LICENSE-2.0)

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.




CHANGELOG.md
# Changelog

## Version 0.9
xx.xx.2023

- Added support for SDT version 4: Conversion from SDT3 to SDT4, markdown, and oneM2M XSD & SVG. 
- Support for namespace prefix in markdown generation.
- Added some short versions for command line arguments.
- Small fixes and improvements. Code restructuring.
- Added support for rich colour output.
- Added new CSV table output of all names, short names, and where they occur in.

## Version 0.8
15.05.2018

- Added *--modelversion** 
- Adopted to new naming scheme for oneM2M XSD and SVG files
- Various fixes for XSD schemas. Support oneM2M Version 2 XSD
- Refactored script directory structure
- Added *--markdownpagebreak* option
- Added first support to generate Swagger files for ModuleClasses and Devices. Not complete yet. 
- Added *-of swagger* option
- Beautified OPML/Mindmanager output
- First work on replacing the hand-crafted output modules by a templating engine (Jinja2, [](http://jinja.pocoo.org/)).
- Profided first templates for markdown and oneM2M XSD.
- Added support for oneM2M release R3
 

## Version 0.7
09.04.2016

- Export to Eclipse Vorto, first version
- First version of export to SVG in oneM2M resource format
- First version of export to oneM2M XSD
- First version of abbreviation support (for oneM2M)
- Fixed errors in Java export
- Fixed errors in OPML export
- Improved markdown export
- Added optional reading of command line arguments from configuration file(s)
- Made SDT3 the default input format
- minor bug fixes


## Version 0.6
04.02.2016

- Improved Java export: output documentation when available, generate static variables for property names, added getter for event payload data points
- Fixed error when exporting SDT2 to markdown
- Added option (``--markdowntables``) to present data points, actions, properties and more in table format

## Version 0.5
13.11.2015

- Added markdown export for SDT3
- Added OPML export for SDT3
- Added support to export SDT3 structure as markdown tables

## Version 0.4
29.10.2015

- Added support for converting SDT2 to SDT3 format
- Added first support for generating Java classes (from SDT3 format only)
- Added --hidedetails option to hide detailed information when generating documentation 

## Version 0.3
03.09.2015

- Fixed duplicate output of DataPoints

## Version 0.2
22.08.2015

- Improved Markdown output for DeviceInfo

## Version 0.1
11.08.2015

- Initial version of SDTTool. 
- Read XML SDT 2.0.1
- Write plain text, markdown and OPML


SDTTool.py
#	SDTTool.py
#
#	Main module for the SDTTool

from xml.etree.ElementTree import XMLParser, ParseError
from sdtv2 import SDT2Parser
from sdtv3 import SDT3Parser
from sdtv4 import SDT4Parser
import common.SDTHelper as SDTHelper, SDTPrinter


import io, sys, traceback, argparse, os
from rich import print

version = '0.9'
description = 'SDTTool ' + version + ' - A tool to read and convert Smart Device Templates.'
epilog = 'Read arguments from one or more configuration files: @file1 @file2 ...|n |n See https://github.com/Homegateway for further information.'


# class LineNumberingParser(XMLParser):
#     def _start_list(self, *args, **kwargs):
#         # Here we assume the default XML parser which is expat
#         # and copy its element position attributes into output Elements
#         element = super(self.__class__, self)._start_list(*args, **kwargs)
#         element._start_line_number = self.parser.CurrentLineNumber
#         element._start_column_number = self.parser.CurrentColumnNumber
#         element._start_byte_index = self.parser.CurrentByteIndex
#         return element

#     def _end(self, *args, **kwargs):
#         element = super(self.__class__, self)._end(*args, **kwargs)
#         element._end_line_number = self.parser.CurrentLineNumber
#         element._end_column_number = self.parser.CurrentColumnNumber
#         element._end_byte_index = self.parser.CurrentByteIndex
#         return element


#
# Parse the data
#
def parseData(target, data):
	"""	Parse the data with the given parser and handle errors.
	"""
	parser = XMLParser(target=target)
	errormsg = ''
	try:
		try:
			parser.feed(data)
		except SyntaxError as err:
			errormsg = str(err)
			print(err)
		except:
			traceback.print_exc()
		finally:
			parser.close()
	except ParseError as err:
		formatted_e = errormsg
		line = int(formatted_e[formatted_e.find("line ") + 5: formatted_e.find(",")])
		column = int(formatted_e[formatted_e.find("column ") + 7:])
		split_str = data.split("\n")
		print(f'{split_str[line - 1]}\n{len(split_str[line - 1][0:column])*"-"}')

	return target.domain, target.nameSpaces


#
#	XML Reader
#
def readSDTXML(inFile, sdtVersion=2):
	""" Read and parse an SDT version 2, 3 or 4 XML.
	"""
	# read the data from the input file
	with open(inFile, 'r') as inputFile:
		data = inputFile.read()

	# select parser
	parser = [ None, None, SDT2Parser, SDT3Parser, SDT4Parser ][sdtVersion]()	# no parsers for v 0 and 1

	# Parse the data
	return parseData(parser, data)


#
#	Output
#
def outputResult(outFile, result):
	"""	Print the output to stdout or to a file.
	"""
	if result == None:
		return
	if outFile == None:
		print(result)
	else:
		path = os.path.dirname(outFile)
		if len(path) > 0:
			os.makedirs(path, exist_ok=True)
		try:
			with open(outFile, 'w') as outputFile:
				outputFile.write(result)
		except IOError as err:
			print(err)


def checkForNamespace(nameSpaces, checkNameSpace):
	"""	Check whether a name space can be found in a list of nameSpaces.
	"""
	for ns in nameSpaces:
		if (ns.find(checkNameSpace) > -1):
			return True
	return False


def main(argv):
	outFile = None

	# Read command line arguments
	
	parser = argparse.ArgumentParser(description=description, epilog=epilog, fromfile_prefix_chars='@', formatter_class=SDTHelper.MultilineFormatter)
	parser.convert_arg_line_to_args = SDTHelper.convertArgLineToArgs

	parser.add_argument('-o', '--outfile', action='store', dest='outFile', default=None, help='The output file or directory for the result. The default is stdout')
	parser.add_argument('-if', '--inputformat', choices=('sdt2', 'sdt3', 'sdt4'), action='store', dest='inputFormat', default='sdt4', help='The input format to read. The default is sdt4')
	parser.add_argument('-of', '--outputformat', choices=('plain', 'opml', 'markdown', 'sdt3', 'sdt4', 'java', 'vorto-dsl', 'onem2m-svg', 'onem2m-xsd', 'swagger', 'apjson'), action='store', dest='outputFormat', default='markdown', help='The output format for the result. The default is markdown')
	parser.add_argument('--hidedetails',  action='store_true', help='Hide the details of module classes and devices when printing documentation')
	parser.add_argument('--markdowntables',  action='store_true', help='Format markdown output as tables for markdown')
	parser.add_argument('--markdownpagebreak',  action='store_true', help='Insert page breaks before ModuleClasse and Device definitions.')
	parser.add_argument('-lf', '--licensefile',  action='store', dest='licensefile', help='Add the text of license file to output files')

	oneM2MArgs = parser.add_argument_group('oneM2M sepcific')
	oneM2MArgs.add_argument('--domain',  action='store', dest='domain', help='Set the domain for the model')
	oneM2MArgs.add_argument('-ns', '--namespaceprefix',  action='store', dest='namespaceprefix', default='m2m', help='Specify the name space prefix for the model')
	oneM2MArgs.add_argument('--abbreviationsinfile', '-abif',  action='store', dest='abbreviationsinfile', help='Specify the file that contains a CSV table of alreadys existing abbreviations.')
	oneM2MArgs.add_argument('--abbreviationlength',  action='store', dest='abbreviationlength', type=int, default=5, help='Specify the maximum length for abbreviations. The default is 5.')
	oneM2MArgs.add_argument('--xsdtargetnamespace',  action='store', dest='xsdtargetnamespace', help='Specify the target namespace for the oneM2M XSD (a URI).')
	oneM2MArgs.add_argument('--modelversion', '-mv', action='store', dest='modelversion', help='Specify the version of the model.')
	oneM2MArgs.add_argument('--svg-with-attributes',  action='store_true', dest='svgwithattributes', help='Generate SVG for ModuleClass attributes as well.')
	oneM2MArgs.add_argument('--xsdnamespacemapping', '-xsdnsmap',  action='store', dest='xsdnamespacemapping', nargs='*', help='Specify the target namespace for the oneM2M XSD (a URI).')
	oneM2MArgs.add_argument('--cdtversion',  action='store', dest='cdtversion', help='Specify the version number of the oneM2M XSD.')

	requiredNamed = parser.add_argument_group('required arguments')
	requiredNamed.add_argument('-i', '--infile', action='store', dest='inFile', required=True, help='The SDT input file to parse')
	
	if len(sys.argv)==1:
		parser.print_help()
		sys.exit(1)

	args 		= parser.parse_args()
	inFile 		= args.inFile
	outFile 	= args.outFile
	inputFormat	= args.inputFormat
	
	moreOptions = {}
	moreOptions['hideDetails'] 					= args.hidedetails
	moreOptions['markdowntables'] 				= args.markdowntables
	moreOptions['pageBreakBeforeMCandDevices'] 	= args.markdownpagebreak
	moreOptions['markdownPageBreak']		 	= args.markdownpagebreak # renamed, therefore twice
	moreOptions['licensefile'] 					= args.licensefile
	moreOptions['domain'] 						= args.domain
	moreOptions['namespaceprefix'] 				= args.namespaceprefix
	moreOptions['abbreviationsinfile'] 			= args.abbreviationsinfile
	moreOptions['abbreviationlength'] 			= args.abbreviationlength
	moreOptions['xsdtargetnamespace'] 			= args.xsdtargetnamespace
	moreOptions['xsdnamespacemapping']			= args.xsdnamespacemapping
	moreOptions['modelversion'] 				= args.modelversion
	moreOptions['outputFormat']					= args.outputFormat
	moreOptions['svgwithattributes']			= args.svgwithattributes
	moreOptions['cdtversion']					= args.cdtversion


	# Read input file. Check for correct format

	if inputFormat == 'sdt2':
		domain, nameSpaces = readSDTXML(inFile, 2)
		if not checkForNamespace(nameSpaces, 'http://homegatewayinitiative.org/xml/dal/2.0'):
			print('[red]ERROR: Namespace "http://homegatewayinitiative.org/xml/dal/2.0" not found in input file.')
			return

	elif inputFormat == 'sdt3':
		domain, nameSpaces = readSDTXML(inFile, 3)
		if not checkForNamespace(nameSpaces, 'http://homegatewayinitiative.org/xml/dal/3.0'):
			print('[red]ERROR: Namespace "http://homegatewayinitiative.org/xml/dal/3.0" not found in input file.')
			return

	elif inputFormat == 'sdt4':
		domain, nameSpaces = readSDTXML(inFile, 4)
		if not checkForNamespace(nameSpaces, 'http://www.onem2m.org/xml/sdt/4.0'):
			print('[red]ERROR: Namespace "http://www.onem2m.org/xml/sdt/4.0" not found in input file.')
			return

	# Output to destination format
	if   args.outputFormat == 'plain':		outputResult(outFile, SDTPrinter.printPlain(domain, moreOptions))
	elif args.outputFormat == 'opml':		outputResult(outFile, SDTPrinter.printOPML(domain, moreOptions))
	elif args.outputFormat == 'markdown':	outputResult(outFile, SDTPrinter.printMarkdown(domain, moreOptions))
	elif args.outputFormat == 'sdt3':		outputResult(outFile, SDTPrinter.printSDT3(domain, inputFormat, moreOptions))
	elif args.outputFormat == 'sdt4':		outputResult(outFile, SDTPrinter.printSDT4(domain, inputFormat, moreOptions))
	elif args.outputFormat == 'java':		SDTPrinter.printJava(domain, inputFormat, outFile, moreOptions)
	elif args.outputFormat == 'vorto-dsl':	SDTPrinter.printVortoDSL(domain, inputFormat, outFile, moreOptions)
	elif args.outputFormat == 'onem2m-svg':	SDTPrinter.printOneM2MSVG(domain, inputFormat, outFile, moreOptions)
	elif args.outputFormat == 'onem2m-xsd':	SDTPrinter.printOneM2MXSD(domain, inputFormat, outFile, moreOptions)
	elif args.outputFormat == 'apjson':		outputResult(outFile, SDTPrinter.printApJSON(domain, inputFormat, outFile, moreOptions))
	elif args.outputFormat == 'swagger':	SDTPrinter.printSwagger(domain, inputFormat, outFile, moreOptions)


if __name__ == "__main__":
	main(sys.argv[1:])
